src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,275 @@
+ * @return {!./layout-rect.LayoutRectDef}
+ */
+function LayoutRectFromDomRect(rect) {
+  return {
+    left: rect.left,
+    top: rect.top,
+    width: rect.width,
+    height: rect.height,
+    bottom: rect.bottom,
+    right: rect.right,
+  };
+}
+
+
+/**
+ * Returns the ratio of the smaller box's area to the larger box's area.
+ * @param {!./layout-rect.LayoutRectDef} smaller
+ * @param {!./layout-rect.LayoutRectDef} larger
+ * @return {number}
+ */
+function intersectionRatio(smaller, larger) {
+  return (smaller.width * smaller.height) / (larger.width * larger.height);
+}
+
+/**
+ * The IntersectionObserverPolyfill class lets any element receive its
+ * intersection data with the viewport. The IntersectionObserverPolyfill acts
+ * like native browser supported IntersectionObserver.
+ * IntersectionObserverPolyfill will
+ * create a PositionObserver class (Not implemented yet) to track the
+ * intersection between element and viewport.
+ * The IntersectionObserver receives a callback function and an optional option
+ * as params. If option is not provided, the default thresholds value will be
+ * used. Whenever the element intersection ratio cross a threshold value,
+ * IntersectionObserverPolyfill will call the callback function.
+ * Note: To share the IntersectionObserverEntry to an iframe, whoever create the
+ * IntersectionObserverPolyfill needs to handle postMessageApi and takes care of
+ * sending messages rate etc outside the IntersectionObserverPolyfill.
+ */
+export class IntersectionObserverPolyfill {
+  /**
+   * @param {!function()} callback.
+   * @param {Object=} opt_option
+   */
+  constructor(callback, opt_option) {
+    /** @private @const {function()} */
+    this.callback_ = callback;
+
+    /**
+     * A list of threshold, sorted in increasing numeric order
+     * @private @const {!Array}
+     */
+    this.threshold_ = opt_option ? opt_option.threshold : DEFAULT_THRESHOLD;
+
+    /**
+     * TODO: Need to support multiple observed elements.
+     * If support multiple elements, will they share the callback and threshold
+     * @private {?Element}
+     */
+    this.element_ = null;
+
+    /** @private {boolean} */
+    this.isAmpElement_ = false;
+
+    /**
+     * A cached value to store getBoundingClientRect() value
+     * if element is not an AMP element
+     * @private {?DOMRect}
+     */
+    this.elementRectForNonAMP_ = null;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: create the PositionObserver class, or create listener for position
+    // info if inside an iframe
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * @param {!Element} element
```

Comment:
Please add doc to explain that it only allow to observe one element.
And add a TODO as well.